File: js/src/observability/health.js
/**
 * Health check module for links-queue.
 *
 * Provides liveness and readiness endpoints with component health details,
 * compatible with Kubernetes health probes.
 *
 * @module observability/health
 *
 * @see REQUIREMENTS.md - REQ-OBS-020 through REQ-OBS-022
 */

/**
 * Health status values.
 */
export const HealthStatus = Object.freeze({
  HEALTHY: 'healthy',
  UNHEALTHY: 'unhealthy',
  DEGRADED: 'degraded',
  UNKNOWN: 'unknown',
});

/**
 * Component health check result.
 *
 * @typedef {Object} ComponentHealth
 * @property {string} status - Health status
 * @property {number} [latency] - Response latency in ms
 * @property {string} [message] - Additional message
 * @property {Object} [details] - Additional details
 */

/**
 * Overall health check result.
 *
 * @typedef {Object} HealthCheckResult
 * @property {string} status - Overall health status
 * @property {number} timestamp - Check timestamp
 * @property {string} version - Application version
 * @property {number} uptime - Uptime in milliseconds
 * @property {Object<string, ComponentHealth>} components - Component health
 */

/**
 * A component health checker.
 */
export class ComponentChecker {
  /**
   * Creates a new ComponentChecker.
   *
   * @param {Object} options - Configuration options
   * @param {string} options.name - Component name
   * @param {Function} options.check - Health check function (async, returns ComponentHealth)
   * @param {boolean} [options.critical=true] - Whether this component is critical for overall health
   * @param {number} [options.timeout=5000] - Check timeout in ms
   */
  constructor(options) {
    this._name = options.name;
    this._check = options.check;
    this._critical = options.critical !== false;
    this._timeout = options.timeout ?? 5000;
  }

  /**
   * Gets the component name.
   *
   * @returns {string} Component name
   */
  get name() {
    return this._name;
  }

  /**
   * Gets whether this component is critical.
   *
   * @returns {boolean} True if critical
   */
  get critical() {
    return this._critical;
  }

  /**
   * Runs the health check.
   *
   * @returns {Promise<ComponentHealth>} Health check result
   */
  async check() {
    const start = Date.now();
    try {
      const result = await Promise.race([
        this._check(),
        new Promise((_, reject) =>
          setTimeout(
            () => reject(new Error('Health check timeout')),
            this._timeout
          )
        ),
      ]);

      return {
        status: result.status ?? HealthStatus.HEALTHY,
        latency: Date.now() - start,
        ...result,
      };
    } catch (error) {
      return {
        status: HealthStatus.UNHEALTHY,
        latency: Date.now() - start,
        message: error.message,
      };
    }
  }
}

/**
 * Health checker for links-queue.
 *
 * Provides liveness, readiness, and detailed health checks
 * compatible with Kubernetes health probes.
 */
export class HealthChecker {
  /**
   * Creates a new HealthChecker.
   *
   * @param {Object} [options] - Configuration options
   * @param {string} [options.version] - Application version
   * @param {Object<string, ComponentChecker>} [options.components] - Pre-registered components
   */
  constructor(options = {}) {
    this._version = options.version ?? '0.0.0';
    this._startTime = Date.now();
    /** @type {Map<string, ComponentChecker>} */
    this._components = new Map();

    // Pre-register components if provided
    if (options.components) {
      for (const [name, checker] of Object.entries(options.components)) {
        this._components.set(name, checker);
      }
    }

    // Internal state
    this._ready = false;
    this._alive = true;
  }

  /**
   * Gets the uptime in milliseconds.
   *
   * @returns {number} Uptime
   */
  get uptime() {
    return Date.now() - this._startTime;
  }

  /**
   * Sets the liveness state.
   *
   * @param {boolean} alive - Whether the service is alive
   */
  setAlive(alive) {
    this._alive = alive;
  }

  /**
   * Sets the readiness state.
   *
   * @param {boolean} ready - Whether the service is ready
   */
  setReady(ready) {
    this._ready = ready;
  }

  /**
   * Registers a component health checker.
   *
   * @param {string} name - Component name
   * @param {Function|ComponentChecker} checker - Health check function or ComponentChecker
   * @param {Object} [options] - Options for function-style checkers
   */
  registerComponent(name, checker, options = {}) {
    if (checker instanceof ComponentChecker) {
      this._components.set(name, checker);
    } else {
      this._components.set(
        name,
        new ComponentChecker({
          name,
          check: checker,
          ...options,
        })
      );
    }
  }

  /**
   * Unregisters a component health checker.
   *
   * @param {string} name - Component name
   * @returns {boolean} True if removed
   */
  unregisterComponent(name) {
    return this._components.delete(name);
  }

  /**
   * Checks liveness.
   * A service is alive if the process is running and not deadlocked.
   *
   * @returns {Promise<{status: string, timestamp: number}>} Liveness result
   */
  async isAlive() {
    // Async for API consistency with other health check methods
    return Promise.resolve({
      status: this._alive ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY,
      timestamp: Date.now(),
    });
  }

  /**
   * Checks readiness.
   * A service is ready if it can accept requests.
   *
   * @returns {Promise<{status: string, timestamp: number}>} Readiness result
   */
  async isReady() {
    // If explicit readiness is set, use it
    if (!this._ready) {
      return {
        status: HealthStatus.UNHEALTHY,
        timestamp: Date.now(),
        message: 'Service not ready',
      };
    }

    // Check critical components
    for (const [name, checker] of this._components) {
      if (checker.critical) {
        const result = await checker.check();
        if (result.status === HealthStatus.UNHEALTHY) {
          return {
            status: HealthStatus.UNHEALTHY,
            timestamp: Date.now(),
            message: `Critical component ${name} is unhealthy`,
            component: name,
          };
        }
      }
    }

    return {
      status: HealthStatus.HEALTHY,
      timestamp: Date.now(),
    };
  }

  /**
   * Gets detailed health status for all components.
   *
   * @returns {Promise<HealthCheckResult>} Detailed health result
   */
  async getHealthDetails() {
    const components = {};
    let overallStatus = HealthStatus.HEALTHY;
    let hasDegraded = false;

    // Check all components in parallel
    const checks = await Promise.all(
      Array.from(this._components.entries()).map(async ([name, checker]) => {
        const result = await checker.check();
        return { name, result, critical: checker.critical };
      })
    );

    for (const { name, result, critical } of checks) {
      components[name] = result;

      if (result.status === HealthStatus.UNHEALTHY) {
        if (critical) {
          overallStatus = HealthStatus.UNHEALTHY;
        } else {
          hasDegraded = true;
        }
      } else if (result.status === HealthStatus.DEGRADED) {
        hasDegraded = true;
      }
    }

    // If not unhealthy but has degraded components, set to degraded
    if (overallStatus === HealthStatus.HEALTHY && hasDegraded) {
      overallStatus = HealthStatus.DEGRADED;
    }

    // Check basic liveness/readiness
    if (!this._alive) {
      overallStatus = HealthStatus.UNHEALTHY;
    } else if (!this._ready && overallStatus === HealthStatus.HEALTHY) {
      overallStatus = HealthStatus.DEGRADED;
    }

    return {
      status: overallStatus,
      timestamp: Date.now(),
      version: this._version,
      uptime: this.uptime,
      ready: this._ready,
      alive: this._alive,
      components,
    };
  }

  /**
   * Registers a queue backend health check.
   *
   * @param {string} name - Backend name
   * @param {Object} backend - Backend instance with isConnected() method
   * @param {Object} [options] - Check options
   */
  registerBackend(name, backend, options = {}) {
    this.registerComponent(
      name,
      () => {
        const isConnected =
          typeof backend.isConnected === 'function'
            ? backend.isConnected()
            : true;

        if (!isConnected) {
          return Promise.resolve({
            status: HealthStatus.UNHEALTHY,
            message: 'Backend disconnected',
          });
        }

        // Try to get stats if available
        if (typeof backend.getStats === 'function') {
          const stats = backend.getStats();
          return Promise.resolve({
            status: HealthStatus.HEALTHY,
            details: {
              connections: stats.connections ?? 1,
              operations: stats.operations ?? {},
            },
          });
        }

        return Promise.resolve({ status: HealthStatus.HEALTHY });
      },
      { critical: true, ...options }
    );
  }

  /**
   * Registers a cluster node health check.
   *
   * @param {string} name - Cluster name
   * @param {Object} cluster - Cluster coordinator instance
   * @param {Object} [options] - Check options
   */
  registerCluster(name, cluster, options = {}) {
    this.registerComponent(
      name,
      () => {
        if (typeof cluster.getStats === 'function') {
          const stats = cluster.getStats();
          const nodeCount = stats.nodeCount ?? 0;

          if (nodeCount === 0) {
            return Promise.resolve({
              status: HealthStatus.UNHEALTHY,
              message: 'No cluster nodes',
            });
          }

          return Promise.resolve({
            status: HealthStatus.HEALTHY,
            details: {
              nodes: nodeCount,
              leader: stats.leader ?? null,
            },
          });
        }

        return Promise.resolve({ status: HealthStatus.UNKNOWN });
      },
      { critical: false, ...options }
    );
  }

  /**
   * Registers a queue health check.
   *
   * @param {string} name - Queue name
   * @param {Object} queue - Queue instance
   * @param {Object} [options] - Check options
   */
  registerQueue(name, queue, options = {}) {
    this.registerComponent(
      name,
      () => {
        if (typeof queue.getStats === 'function') {
          const stats = queue.getStats();
          return Promise.resolve({
            status: HealthStatus.HEALTHY,
            details: {
              depth: stats.depth ?? 0,
              inFlight: stats.inFlight ?? 0,
            },
          });
        }

        return Promise.resolve({ status: HealthStatus.HEALTHY });
      },
      { critical: false, ...options }
    );
  }
}

/**
 * Express/Fastify compatible middleware for health endpoints.
 *
 * @param {Object} options - Middleware options
 * @param {HealthChecker} options.checker - Health checker instance
 * @param {string} [options.livePath='/health/live'] - Liveness endpoint path
 * @param {string} [options.readyPath='/health/ready'] - Readiness endpoint path
 * @param {string} [options.detailsPath='/health/details'] - Details endpoint path
 * @returns {Function} Middleware function
 */
export const healthMiddleware = (options) => {
  const { checker } = options;
  const livePath = options.livePath ?? '/health/live';
  const readyPath = options.readyPath ?? '/health/ready';
  const detailsPath = options.detailsPath ?? '/health/details';

  return async (req, res, next) => {
    // Support both Express-style (req.path) and raw HTTP (req.url)
    const requestPath =
      req.path ?? new URL(req.url, 'http://localhost').pathname;

    const setJsonResponse = (statusCode, body) => {
      res.statusCode = statusCode;
      res.setHeader('Content-Type', 'application/json');
      res.end(JSON.stringify(body));
    };

    if (requestPath === livePath) {
      const result = await checker.isAlive();
      setJsonResponse(
        result.status === HealthStatus.HEALTHY ? 200 : 503,
        result
      );
      return;
    }

    if (requestPath === readyPath) {
      const result = await checker.isReady();
      setJsonResponse(
        result.status === HealthStatus.HEALTHY ? 200 : 503,
        result
      );
      return;
    }

    if (requestPath === detailsPath) {
      const result = await checker.getHealthDetails();
      let statusCode = 200;
      if (result.status === HealthStatus.UNHEALTHY) {
        statusCode = 503;
      } else if (result.status === HealthStatus.DEGRADED) {
        statusCode = 200; // Degraded is still operational
      }
      setJsonResponse(statusCode, result);
      return;
    }

    if (typeof next === 'function') {
      next();
    }
  };
};

/**
 * Creates a health checker pre-configured for a queue server.
 *
 * @param {Object} options - Configuration options
 * @param {string} [options.version] - Application version
 * @param {Object} [options.backend] - Backend instance to monitor
 * @param {Object} [options.cluster] - Cluster coordinator to monitor
 * @param {Object<string, Object>} [options.queues] - Queues to monitor
 * @returns {HealthChecker} Configured health checker
 */
export const createServerHealthChecker = (options = {}) => {
  const checker = new HealthChecker({
    version: options.version,
  });

  // Register backend if provided
  if (options.backend) {
    checker.registerBackend('backend', options.backend);
  }

  // Register cluster if provided
  if (options.cluster) {
    checker.registerCluster('cluster', options.cluster);
  }

  // Register queues if provided
  if (options.queues) {
    for (const [name, queue] of Object.entries(options.queues)) {
      checker.registerQueue(`queue_${name}`, queue);
    }
  }

  return checker;
};

File: rust/src/observability/health.rs
//! Health check module for links-queue.
//!
//! Provides liveness and readiness endpoints with component health details,
//! compatible with Kubernetes health probes.

use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

/// Health status values.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum HealthStatus {
    Healthy,
    Unhealthy,
    Degraded,
    Unknown,
}

impl HealthStatus {
    /// Returns the string representation.
    #[must_use]
    pub const fn as_str(&self) -> &'static str {
        match self {
            Self::Healthy => "healthy",
            Self::Unhealthy => "unhealthy",
            Self::Degraded => "degraded",
            Self::Unknown => "unknown",
        }
    }
}

impl std::fmt::Display for HealthStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl Default for HealthStatus {
    fn default() -> Self {
        Self::Unknown
    }
}

/// Component health check result.
#[derive(Debug, Clone)]
pub struct ComponentHealth {
    /// Health status.
    pub status: HealthStatus,
    /// Response latency in milliseconds.
    pub latency: Option<u64>,
    /// Additional message.
    pub message: Option<String>,
    /// Additional details.
    pub details: HashMap<String, serde_json::Value>,
}

impl ComponentHealth {
    /// Creates a healthy result.
    #[must_use]
    pub fn healthy() -> Self {
        Self {
            status: HealthStatus::Healthy,
            latency: None,
            message: None,
            details: HashMap::new(),
        }
    }

    /// Creates an unhealthy result.
    #[must_use]
    pub fn unhealthy(message: impl Into<String>) -> Self {
        Self {
            status: HealthStatus::Unhealthy,
            latency: None,
            message: Some(message.into()),
            details: HashMap::new(),
        }
    }

    /// Creates a degraded result.
    #[must_use]
    pub fn degraded(message: impl Into<String>) -> Self {
        Self {
            status: HealthStatus::Degraded,
            latency: None,
            message: Some(message.into()),
            details: HashMap::new(),
        }
    }

    /// Sets the latency.
    #[must_use]
    pub const fn with_latency(mut self, latency_ms: u64) -> Self {
        self.latency = Some(latency_ms);
        self
    }

    /// Adds a detail field.
    #[must_use]
    pub fn with_detail(
        mut self,
        key: impl Into<String>,
        value: impl Into<serde_json::Value>,
    ) -> Self {
        self.details.insert(key.into(), value.into());
        self
    }

    /// Converts to JSON.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        let mut obj = serde_json::Map::new();
        obj.insert(
            "status".to_string(),
            serde_json::json!(self.status.as_str()),
        );

        if let Some(latency) = self.latency {
            obj.insert("latency".to_string(), serde_json::json!(latency));
        }
        if let Some(ref msg) = self.message {
            obj.insert("message".to_string(), serde_json::json!(msg));
        }
        if !self.details.is_empty() {
            obj.insert("details".to_string(), serde_json::json!(self.details));
        }

        serde_json::Value::Object(obj)
    }
}

/// Overall health check result.
#[derive(Debug, Clone)]
pub struct HealthCheckResult {
    /// Overall health status.
    pub status: HealthStatus,
    /// Check timestamp (milliseconds since Unix epoch).
    pub timestamp: u64,
    /// Application version.
    pub version: String,
    /// Uptime in milliseconds.
    pub uptime: u64,
    /// Whether the service is ready.
    pub ready: bool,
    /// Whether the service is alive.
    pub alive: bool,
    /// Component health results.
    pub components: HashMap<String, ComponentHealth>,
}

impl HealthCheckResult {
    /// Converts to JSON.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        let mut components = serde_json::Map::new();
        for (name, health) in &self.components {
            components.insert(name.clone(), health.to_json());
        }

        serde_json::json!({
            "status": self.status.as_str(),
            "timestamp": self.timestamp,
            "version": self.version,
            "uptime": self.uptime,
            "ready": self.ready,
            "alive": self.alive,
            "components": components
        })
    }
}

/// Type alias for health check function.
pub type HealthCheckFn =
    Box<dyn Fn() -> Pin<Box<dyn Future<Output = ComponentHealth> + Send>> + Send + Sync>;

/// A component health checker.
pub struct ComponentChecker {
    name: String,
    check: HealthCheckFn,
    critical: bool,
    timeout: Duration,
}

impl ComponentChecker {
    /// Creates a new ComponentChecker.
    pub fn new<F, Fut>(name: impl Into<String>, check: F) -> Self
    where
        F: Fn() -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ComponentHealth> + Send + 'static,
    {
        Self {
            name: name.into(),
            check: Box::new(move || Box::pin(check())),
            critical: true,
            timeout: Duration::from_secs(5),
        }
    }

    /// Sets whether this component is critical.
    #[must_use]
    pub fn with_critical(mut self, critical: bool) -> Self {
        self.critical = critical;
        self
    }

    /// Sets the check timeout.
    #[must_use]
    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.timeout = timeout;
        self
    }

    /// Gets the component name.
    #[must_use]
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Gets whether this component is critical.
    #[must_use]
    pub fn is_critical(&self) -> bool {
        self.critical
    }

    /// Runs the health check.
    pub async fn check(&self) -> ComponentHealth {
        let start = Instant::now();

        let check_future = (self.check)();

        // Use tokio timeout
        match tokio::time::timeout(self.timeout, check_future).await {
            Ok(result) => result.with_latency(start.elapsed().as_millis() as u64),
            Err(_) => ComponentHealth::unhealthy("Health check timeout")
                .with_latency(start.elapsed().as_millis() as u64),
        }
    }
}

/// Health checker for links-queue.
pub struct HealthChecker {
    version: String,
    start_time: Instant,
    components: Arc<RwLock<HashMap<String, ComponentChecker>>>,
    ready: Arc<RwLock<bool>>,
    alive: Arc<RwLock<bool>>,
}

impl HealthChecker {
    /// Creates a new HealthChecker.
    #[must_use]
    pub fn new(version: impl Into<String>) -> Self {
        Self {
            version: version.into(),
            start_time: Instant::now(),
            components: Arc::new(RwLock::new(HashMap::new())),
            ready: Arc::new(RwLock::new(false)),
            alive: Arc::new(RwLock::new(true)),
        }
    }

    /// Gets the uptime in milliseconds.
    #[must_use]
    pub fn uptime(&self) -> u64 {
        self.start_time.elapsed().as_millis() as u64
    }

    /// Sets the liveness state.
    pub fn set_alive(&self, alive: bool) {
        if let Ok(mut a) = self.alive.write() {
            *a = alive;
        }
    }

    /// Sets the readiness state.
    pub fn set_ready(&self, ready: bool) {
        if let Ok(mut r) = self.ready.write() {
            *r = ready;
        }
    }

    /// Gets the alive state.
    #[must_use]
    pub fn is_alive(&self) -> bool {
        self.alive.read().map(|a| *a).unwrap_or(false)
    }

    /// Gets the ready state.
    #[must_use]
    pub fn is_ready(&self) -> bool {
        self.ready.read().map(|r| *r).unwrap_or(false)
    }

    /// Registers a component health checker.
    pub fn register_component(&self, checker: ComponentChecker) {
        if let Ok(mut components) = self.components.write() {
            components.insert(checker.name.clone(), checker);
        }
    }

    /// Unregisters a component.
    pub fn unregister_component(&self, name: &str) -> bool {
        if let Ok(mut components) = self.components.write() {
            components.remove(name).is_some()
        } else {
            false
        }
    }

    /// Checks liveness.
    #[must_use]
    pub fn check_alive(&self) -> LivenessResult {
        let alive = self.alive.read().map(|a| *a).unwrap_or(false);
        LivenessResult {
            status: if alive {
                HealthStatus::Healthy
            } else {
                HealthStatus::Unhealthy
            },
            timestamp: current_timestamp(),
        }
    }

    /// Checks readiness.
    pub async fn check_ready(&self) -> ReadinessResult {
        let ready = self.ready.read().map(|r| *r).unwrap_or(false);

        if !ready {
            return ReadinessResult {
                status: HealthStatus::Unhealthy,
                timestamp: current_timestamp(),
                message: Some("Service not ready".to_string()),
                component: None,
            };
        }

        // Check critical components
        let components = self.components.read().ok();
        if let Some(components) = components {
            for (name, checker) in components.iter() {
                if checker.is_critical() {
                    let result = checker.check().await;
                    if result.status == HealthStatus::Unhealthy {
                        return ReadinessResult {
                            status: HealthStatus::Unhealthy,
                            timestamp: current_timestamp(),
                            message: Some(format!("Critical component {} is unhealthy", name)),
                            component: Some(name.clone()),
                        };
                    }
                }
            }
        }

        ReadinessResult {
            status: HealthStatus::Healthy,
            timestamp: current_timestamp(),
            message: None,
            component: None,
        }
    }

    /// Gets detailed health status for all components.
    pub async fn get_health_details(&self) -> HealthCheckResult {
        let mut component_results = HashMap::new();
        let mut overall_status = HealthStatus::Healthy;
        let mut has_degraded = false;

        // Check all components
        let components = self.components.read().ok();
        if let Some(components) = components {
            for (name, checker) in components.iter() {
                let result = checker.check().await;

                if result.status == HealthStatus::Unhealthy {
                    if checker.is_critical() {
                        overall_status = HealthStatus::Unhealthy;
                    } else {
                        has_degraded = true;
                    }
                } else if result.status == HealthStatus::Degraded {
                    has_degraded = true;
                }

                component_results.insert(name.clone(), result);
            }
        }

        // If not unhealthy but has degraded components
        if overall_status == HealthStatus::Healthy && has_degraded {
            overall_status = HealthStatus::Degraded;
        }

        // Check basic liveness/readiness
        let alive = self.alive.read().map(|a| *a).unwrap_or(false);
        let ready = self.ready.read().map(|r| *r).unwrap_or(false);

        if !alive {
            overall_status = HealthStatus::Unhealthy;
        } else if !ready && overall_status == HealthStatus::Healthy {
            overall_status = HealthStatus::Degraded;
        }

        HealthCheckResult {
            status: overall_status,
            timestamp: current_timestamp(),
            version: self.version.clone(),
            uptime: self.uptime(),
            ready,
            alive,
            components: component_results,
        }
    }

    /// Registers a backend health check.
    pub fn register_backend<F, Fut>(&self, name: &str, check: F, critical: bool)
    where
        F: Fn() -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ComponentHealth> + Send + 'static,
    {
        let checker = ComponentChecker::new(name, check).with_critical(critical);
        self.register_component(checker);
    }

    /// Registers a simple sync check as a component.
    pub fn register_sync_check<F>(&self, name: &str, check: F, critical: bool)
    where
        F: Fn() -> ComponentHealth + Send + Sync + 'static,
    {
        let check = Arc::new(check);
        let checker = ComponentChecker::new(name, move || {
            let check = Arc::clone(&check);
            async move { check() }
        })
        .with_critical(critical);
        self.register_component(checker);
    }
}

impl Default for HealthChecker {
    fn default() -> Self {
        Self::new("0.0.0")
    }
}

impl Clone for HealthChecker {
    fn clone(&self) -> Self {
        Self {
            version: self.version.clone(),
            start_time: self.start_time,
            components: Arc::clone(&self.components),
            ready: Arc::clone(&self.ready),
            alive: Arc::clone(&self.alive),
        }
    }
}

/// Liveness check result.
#[derive(Debug, Clone)]
pub struct LivenessResult {
    pub status: HealthStatus,
    pub timestamp: u64,
}

impl LivenessResult {
    /// Converts to JSON.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        serde_json::json!({
            "status": self.status.as_str(),
            "timestamp": self.timestamp
        })
    }
}

/// Readiness check result.
#[derive(Debug, Clone)]
pub struct ReadinessResult {
    pub status: HealthStatus,
    pub timestamp: u64,
    pub message: Option<String>,
    pub component: Option<String>,
}

impl ReadinessResult {
    /// Converts to JSON.
    #[must_use]
    pub fn to_json(&self) -> serde_json::Value {
        let mut obj = serde_json::Map::new();
        obj.insert(
            "status".to_string(),
            serde_json::json!(self.status.as_str()),
        );
        obj.insert("timestamp".to_string(), serde_json::json!(self.timestamp));

        if let Some(ref msg) = self.message {
            obj.insert("message".to_string(), serde_json::json!(msg));
        }
        if let Some(ref comp) = self.component {
            obj.insert("component".to_string(), serde_json::json!(comp));
        }

        serde_json::Value::Object(obj)
    }
}

/// Gets the current timestamp in milliseconds.
fn current_timestamp() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_health_status() {
        assert_eq!(HealthStatus::Healthy.as_str(), "healthy");
        assert_eq!(HealthStatus::Unhealthy.as_str(), "unhealthy");
        assert_eq!(HealthStatus::Degraded.as_str(), "degraded");
        assert_eq!(HealthStatus::Unknown.as_str(), "unknown");
    }

    #[test]
    fn test_component_health() {
        let health = ComponentHealth::healthy()
            .with_latency(10)
            .with_detail("connections", 5);

        assert_eq!(health.status, HealthStatus::Healthy);
        assert_eq!(health.latency, Some(10));
        assert!(health.details.contains_key("connections"));
    }

    #[test]
    fn test_component_health_unhealthy() {
        let health = ComponentHealth::unhealthy("Connection failed");

        assert_eq!(health.status, HealthStatus::Unhealthy);
        assert_eq!(health.message, Some("Connection failed".to_string()));
    }

    #[test]
    fn test_health_checker_basic() {
        let checker = HealthChecker::new("1.0.0");

        assert!(!checker.is_ready());
        assert!(checker.is_alive());

        checker.set_ready(true);
        assert!(checker.is_ready());

        checker.set_alive(false);
        assert!(!checker.is_alive());
    }

    #[test]
    fn test_liveness_result() {
        let checker = HealthChecker::new("1.0.0");
        let result = checker.check_alive();

        assert_eq!(result.status, HealthStatus::Healthy);
        assert!(result.timestamp > 0);
    }

    #[test]
    fn test_liveness_result_unhealthy() {
        let checker = HealthChecker::new("1.0.0");
        checker.set_alive(false);
        let result = checker.check_alive();

        assert_eq!(result.status, HealthStatus::Unhealthy);
    }

    #[test]
    fn test_health_check_result_json() {
        let result = HealthCheckResult {
            status: HealthStatus::Healthy,
            timestamp: 1234567890,
            version: "1.0.0".to_string(),
            uptime: 60000,
            ready: true,
            alive: true,
            components: HashMap::new(),
        };

        let json = result.to_json();
        assert_eq!(json["status"], "healthy");
        assert_eq!(json["version"], "1.0.0");
        assert_eq!(json["uptime"], 60000);
    }

    #[test]
    fn test_register_sync_check() {
        let checker = HealthChecker::new("1.0.0");

        checker.register_sync_check("test", || ComponentHealth::healthy(), true);

        let components = checker.components.read().unwrap();
        assert!(components.contains_key("test"));
        assert!(components.get("test").unwrap().is_critical());
    }
}
